apache2/msc_util.c

```diff
@@ -1779,6 +1779,98 @@ int urldecode_nonstrict_inplace_ex(unsigned char *input, long int input_len, int
     return count;
 }
 
+// RFC 5987 allows HTTP values to be non-ASCII characters.
+// As per RFC 5987 the value can be encoded as UTF-8 or ISO-8859-1. Since HTTP doesn't support non-ASCII characters, 
+// when using the extended value format specified by RFC 5987 the value is double encoded.
+// The value is first encoded into an octet stream in the format specified in the extended format (UTF-8 or ISO-8859-1)
+// and then the octet stream itself is encoded using URL encoding.
+// RFC 5987 specifies the extended value format to be as follows:
+//  `value = <utf-8|iso-8859-1>'<language>'<URL encoded octect stream>.
+// The function, takes in a RFC 5987 encoded value and decodes it to a UTF-8 encoded string (even if the RFC 5987 was encoded using ISO-8859-1).
+char *rfc5987_decode(apr_pool_t *mptmp, char *value)
+{
+    const char utf8[] = "utf-8";
+    const char iso88591[] = "iso-8859-1";
+    int len = strlen(value);
+    char *urlEncodedValue = strstr(value, "'");
+
+    if (!urlEncodedValue)
+    {
+        return NULL;
+    }
+
+    // Find the URL encoded string. 'URLEncodedValue' is already set to the first occurence of the quote, the format of the extended value as per RFC 5987 is
+    // `value = <utf-8|iso-8859-1>'<language>'<URL encoded octect stream>
```

Comment:
just to confirm in the <language>, there shouldn't be any "'", right?

Reply:
Didn't follow? The encoded string will have quotes?

Reply:
Can you complete your statement above?